src/itwinai/torch/tuning.py

```diff
@@ -0,0 +1,137 @@
+from ray.tune.schedulers import (
+    AsyncHyperBandScheduler,
+    HyperBandForBOHB,
+    HyperBandScheduler,
+    PopulationBasedTraining,
+)
+from ray.tune.schedulers.pb2 import PB2  # Population Based Bandits
+from ray.tune.search.bayesopt import BayesOptSearch
+from ray.tune.search.bohb import TuneBOHB
+from ray.tune.search.hyperopt import HyperOptSearch
+
+
+def get_raytune_search_alg(
+    tune_config, seeds=False
+) -> TuneBOHB | BayesOptSearch | HyperOptSearch | None:
+    """Get the appropriate Ray Tune search algorithm based on the provided configuration.
+
+    Args:
+        tune_config (Dict): Configuration dictionary specifying the search algorithm,
+            metric, mode, and, depending on the search algorithm, other parameters.
+        seeds (bool, optional): Whether to use a fixed seed for reproducibility for some
+            search algorithms that take a seed. Defaults to False.
+
+    Returns:
+        An instance of the chosen Ray Tune search algorithm or None if no search algorithm is
+            used or if the search algorithm does not match any of the supported options.
+
+    Notes:
+        - `TuneBOHB` is automatically chosen for BOHB scheduling.
+    """
+    scheduler = tune_config.get("scheduler", {}).get("name")
+
+    search_alg = tune_config.get("search_alg", {}).get("name")
+
+    if (scheduler == "pbt") or (scheduler == "pb2"):
+        if search_alg is None:
+            return None
+        else:
+            print(
+                "INFO: Using schedule '{}' \
+                    is not compatible with Ray Tune search algorithms.".format(scheduler)
+            )
+            print(
+                "INFO: Using the Ray Tune '{}' scheduler without search algorithm".format(
+                    scheduler
+                )
+            )
+
+    if (scheduler == "bohb") or (scheduler == "BOHB"):
+        print("INFO: Using TuneBOHB search algorithm since it is required for BOHB shedule")
+        if seeds:
+            seed = 1234
+        else:
+            seed = None
+        return TuneBOHB(
+            seed=seed,
+        )
+
+    # requires pip install bayesian-optimization
+    if search_alg == "bayes":
+        print("INFO: Using BayesOptSearch")
+        return BayesOptSearch(
+            random_search_steps=tune_config["search_alg"]["n_random_steps"],
+        )
+
+    # requires pip install hyperopt
+    if search_alg == "hyperopt":
+        print("INFO: Using HyperOptSearch")
+        return HyperOptSearch(
+            n_initial_points=tune_config["search_alg"]["n_random_steps"],
+            # points_to_evaluate=,
+        )
+    else:
```

Comment:
This "else" confused me for a second since all the previous checks are just "if" checks with no "elif"s. Having the chain like that is fine (since you return under each one), but then by using an else at the end it makes it seem like the else somehow "belongs" to only the last "if", instead of the whole chain. I think I would just remove it, since you're returning in the "if"s anyway.

Reply:
Did this on the other branch